source/agora/crypto/Key.d

```diff
@@ -121,35 +147,51 @@ public struct PublicKey
     /*private*/ Point data;
     alias data this;
 
+    string hrp = HumanReadablePart;
+
     /// Construct an instance from binary data
-    public this (const Point args) pure nothrow @safe @nogc
+    public this (const Point args, const string hrp = HumanReadablePart)
+        nothrow @safe @nogc
     {
         this.data = args;
+
+        auto wkdata = WKKeyData.Keys[this.data];
+        if (wkdata != KeyData.init)
+            this.hrp = wkdata.hrp;
+        else
+            this.hrp = hrp;
     }
```

Comment:
This is not a sound interface. There are two main cases for the second parameter:
- The user provides it, expecting it to be respected;
- The user doesn't provide it, expects a sane default;

The former use case is not covered: If the user provides it, the constructor might still overwrite it.
A much better interface here is to expose two overloads, one that doesn't take a second parameter and do the HRP lookup, and the other that does and just use that.

Reply:
I moved the logic into the `toString` function. Ty.